sled-agent/src/sim/instance.rs

```diff
@@ -47,33 +76,109 @@ impl Simulatable for SimInstance {
+                    // that. This arm abstracts this behavior away and just
+                    // allows a fake instance to transition right back to a
+                    // running state after being stopped.
+                    //
+                    // This will change in the future when the sled agents (both
+                    // real and simulated) split "registering" an instance with
+                    // the agent and actually starting it into separate actions.
+                    self.state.transition(ApiInstanceState::Starting);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Starting);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Running);
                 }
-                InstanceStateRequested::Destroyed => {
-                    PropolisInstanceState::Destroyed
+                ApiInstanceState::Repairing => {
+                    return Err(Error::invalid_request(&format!(
+                        "can't request state Running with terminal state {}",
+                        self.terminal_state()
+                    )))
                 }
-                _ => panic!("Unexpected desired state: {}", desired.run_state),
-            };
-            self.state.observe_transition(&observed)
+            },
+            InstanceStateRequested::Stopped => match self.terminal_state() {
+                ApiInstanceState::Creating => {
+                    self.state.transition(ApiInstanceState::Destroyed);
+                }
+                ApiInstanceState::Running => {
+                    self.state.transition(ApiInstanceState::Stopping);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Stopping);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Stopped);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Destroyed);
+                }
+                // Idempotently allow requests to stop an instance that is
+                // already stopping.
+                ApiInstanceState::Stopping
+                | ApiInstanceState::Stopped
+                | ApiInstanceState::Destroyed => {}
+                _ => {
+                    return Err(Error::invalid_request(&format!(
+                        "can't request state Stopped with terminal state {}",
+                        self.terminal_state()
+                    )))
+                }
+            },
+            InstanceStateRequested::Reboot => match self.terminal_state() {
+                ApiInstanceState::Running => {
+                    self.state.transition(ApiInstanceState::Rebooting);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Rebooting);
+                    self.propolis_queue
+                        .push_back(PropolisInstanceState::Running);
```

Comment:
Is this idempotent? If we reboot while rebooting, will this enqueue another reboot?

Reply:
Currently we will ([relevant Propolis code](https://github.com/oxidecomputer/propolis/blob/master/bin/propolis-server/src/lib/vm/request_queue.rs#L352)), but I could buy that that's not quite the right behavior; it certainly differs from how we handle other state transitions, and I can see why it's not intuitive for things to behave this way. I filed [propolis#363](https://github.com/oxidecomputer/propolis/issues/363) to track over there. I'll fix the simulated behavior up in this PR as well (just to avoid having to remember to do it later).

Reply:
Gotcha. If this matches what we're doing in propolis, I'm fine with it, and having that supplementary issue to follow-up seems like the right next step.